### DISCUSSION TITLE
Moving to TypeGraphQL

### DISCUSSION THREAD

**Discussion by HarunKilic** [vendurehq/vendure#1326] [General]
TypeGraphQL https://typegraphql.com/  has its benefits for deduplicating types, queries and mutations. 
The types will also be typed instead of writing in gql template strings. 

The idea with this discussion is to show proof of consept and/or how we can design this in Vendure. 
Join the journey :)

**Comment by michaelbromley** [comment]
Thanks for starting this discussion. Here are some points I would be interested in finding out more about:

1. Some types in Vendure differ between the Admin API & Shop API. For example, This is handled currently by defining a common `Product` type, and then for the Admin API only, we define additional fields like this: https://github.com/vendure-ecommerce/vendure/blob/820ee684286e64b77e8e0979ead72c410ceb9442/packages/core/src/api/schema/admin-api/product-admin.type.graphql#L1-L4
2. Can it be implemented without breaking changes? If not it could be something to consider for Vendure v2
3. Vendure performs quite a bit of direct schema manipulation at runtime, e.g. to handle the automatic PaginatedList inputs. Would this still be possible?

**Comment by rem4ik4ever** [comment]
Here is a quick POC https://github.com/vendure-ecommerce/vendure/pull/1334. See description of changes inside.

**Reply by michaelbromley** [reply]
Thank you for this! I will make some time soon to dive into your changes and try to understand the implications.
